crates/provider-blobstore-nats/src/blobstore.rs

```diff
@@ -0,0 +1,678 @@
+        }
+        .await
+        .map_err(|err: anyhow::Error| err.to_string()))
+    }
+
+    // Check if a NATS Blobstore Container exists
+    #[instrument(level = "debug", skip(self))]
+    async fn container_exists(
+        &self,
+        context: Option<Context>,
+        name: String,
+    ) -> anyhow::Result<Result<bool, String>> {
+        Ok(async {
+            propagate_trace_for_ctx!(context);
+
+            // Retrieve the NATS blobstore connection
+            let blobstore = self
+                .get_blobstore(context)
+                .await
+                .map_err(|e| e.to_string())?;
+
+            // Check if the container exists
+            match blobstore.jetstream.get_object_store(&name).await {
+                Ok(_) => Ok(true),
+                Err(e)
+                    if matches!(
+                        e.kind(),
+                        async_nats::jetstream::context::ObjectStoreErrorKind::GetStore
+                    ) =>
+                {
+                    Ok(false)
+                }
+                Err(e) => Err(format!("failed to check container existence: {e}")),
+            }
+        }
+        .await
+        .map_err(|err| err.to_string()))
+    }
+
+    /// Retrieve data from an object in the specified NATS blobstore Container
+    /// Optionally specify start and end byte offsets for partial reads
+    #[instrument(level = "debug", skip(self))]
+    async fn get_container_data(
+        &self,
+        context: Option<Context>,
+        id: bindings::wrpc::blobstore::types::ObjectId,
+        start: u64,
+        end: u64,
+    ) -> anyhow::Result<
+        Result<
+            (
+                Pin<Box<dyn Stream<Item = Bytes> + Send>>,
+                Pin<Box<dyn Future<Output = Result<(), String>> + Send>>,
+            ),
+            String,
+        >,
+    > {
+        use tokio::io::AsyncReadExt; // Import the trait to use `read`
+
+        Ok(async {
+            propagate_trace_for_ctx!(context);
+
+            // Retrieve the NATS blobstore connection
+            let blobstore = self
+                .get_blobstore(context)
+                .await
+                .context("failed to get NATS Blobstore connection")?;
+
+            // Get the container (object store) for the specified container name
+            let container = blobstore
+                .jetstream
+                .get_object_store(&id.container)
+                .await
+                .context("failed to get container")?;
+
+            // Retrieve the object data as a stream
+            let mut object = container
+                .get(&id.object)
+                .await
+                .context("failed to get object data")?;
+
+            // Create a channel to stream the data
+            let (tx, rx) = mpsc::channel(16);
+            anyhow::Ok((
+                Box::pin(ReceiverStream::new(rx)) as Pin<Box<dyn Stream<Item = _> + Send>>,
+                Box::pin(async move {
```

Comment:
Can this be simplified to one `async move`?

Reply:
When I was working on the provider, I had tried, but couldn't; [azure-blobstore-provider](https://github.com/wasmCloud/wasmCloud/blob/acdde189149b29732b31055924a97ceb68259d13/crates/provider-blobstore-azure/src/lib.rs#L449) has the same implementation.